syncapi/storage/sqlite3/current_room_state_table.go

```diff
@@ -242,16 +242,16 @@ func (s *currentRoomStateStatements) selectEventsWithEventIDs(
 	return rowsToStreamEvents(rows)
 }
 
-func rowsToEvents(rows *sql.Rows) ([]gomatrixserverlib.Event, error) {
-	result := []gomatrixserverlib.Event{}
+func rowsToEvents(rows *sql.Rows) ([]gomatrixserverlib.HeaderedEvent, error) {
+	result := []gomatrixserverlib.HeaderedEvent{}
 	for rows.Next() {
 		var eventBytes []byte
 		if err := rows.Scan(&eventBytes); err != nil {
 			return nil, err
 		}
 		// TODO: Handle redacted events
-		ev, err := gomatrixserverlib.NewEventFromTrustedJSON(eventBytes, false)
-		if err != nil {
+		var ev gomatrixserverlib.HeaderedEvent
+		if err := json.Unmarshal(eventBytes, &ev); err != nil {
```

Comment:
So I've seen this a lot where we are no longer using `NewEventFromTrustedJSON` but instead unmarshalling directly into a `HeaderedEvent`. What are the implications of this? Does the default `Event.UnmarshalJSON` do it trusted or untrusted? Performance impacts?

Reply:
`Event.UnmarshalJSON` delegates to `NewEventFromUntrustedJSON`. The net result of this is that we will double-check the event hashes, and if the hash doesn't match on the first pass, to run the redaction algorithm and try again. Since events from the outside world will have already been through this path before persistence, the event hashes will be re-checked but redaction won't be done.

There are probably ways that we can optimise this but I'd rather get it working first.